compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,534 @@
+            } else if adt_def.is_struct() && adt_def.repr().transparent() {
+                // <element-type1>
+                let mut tys = adt_def.all_fields().map(|field| field.ty(tcx, substs));
+                let ty0 = tys.next().unwrap();
+                // Generalize any repr(transparent) user-defined type that is either a pointer or
+                // reference, and either references itself or any other type that contains or
+                // references itself, to avoid a reference cycle.
+                if ty0.is_any_ptr() && ty0.contains(ty) {
+                    s.push_str(&encode_ty(
+                        tcx,
+                        ty0,
+                        dict,
+                        options | EncodeTyOptions::GENERALIZE_POINTERS,
+                    ));
+                } else {
+                    s.push_str(&encode_ty(tcx, ty0, dict, options));
+                }
+            } else {
+                // u<length><name>[I<element-type1..element-typeN>E] as vendor extended type
+                s.push_str(&encode_ty_name(tcx, def_id));
+                if adt_def.is_struct() || adt_def.is_union() {
+                    let tys: Vec<Ty<'tcx>> = substs.types().map(|ty| ty).collect();
+                    if !tys.is_empty() {
+                        s.push('I');
+                        for ty in tys {
+                            s.push_str(&encode_ty(tcx, ty, dict, options));
+                        }
+                        s.push('E');
+                    }
+                }
+                s = format!("u{}{}", s.len(), &s);
+                compress(dict, (ty, TyQ::None), &mut s);
+            }
+            typeid.push_str(&s);
+        }
+
+        ty::Foreign(def_id) => {
+            // <length><name>, where <name> is <unscoped-name>
+            let mut s = String::new();
+            let name = format!("{}", tcx.item_name(*def_id));
+            s.push_str(&format!("{}", name.len()));
+            s.push_str(&name);
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Function types
+        ty::FnDef(def_id, ..) => {
+            // u2fnI<length><name>E as vendor extended type
+            let mut s = String::from("u2fnI");
+            s.push_str(&encode_ty_name(tcx, *def_id));
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Closure(_, substs) | ty::Generator(_, substs, _) => {
+            // u7closureI<return-type><element-type1..element-typeN>E as vendor extended type
+            let mut s = String::from("u7closureI");
+            let sig = substs.as_closure().sig();
+            // Push str without the initial 'F' but with the closing 'E'
+            s.push_str(&(encode_fnsig(tcx, &sig.skip_binder(), dict)[1..]));
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
```

Comment:
This should be encoded in terms of `def_id` and `substs` instead.

Reply:
Humm, wouldn't it break type membership for functions that receive closures of a certain type as parameters?

Reply:
Those are distinct types, and shouldn't be grouped together merely because they have the same signature.

Reply:
Don't functions that receive closures expect closures of the same type by specifying their signatures only?

Reply:
Each closure expression has its own unique type.

Reply:
This is where this encoding's implementation's concern differ from symbol mangling for linking. I'm concerned if two functions that receive the same type of closure would be compatible. Sorry for my lack of Rust knowledge, but as far as I'm reading on the documentation and experimenting with, it seems that functions that receive closures expect closures of the same type by specifying their signatures only, right?

Reply:
Humm, it seems closures as parameters are encoded as `ty::Dynamic`, so the encoding here would just group the closures themselves as destinations. Am I correct?

Reply:
Closure expression have their own unique type (represented by `ty::Closure`). This type will implement some of `Fn*` traits, and consequently closures can be turned into corresponding trait objects `dyn Fn*` (represented by `ty::Dynamic`).

```rust
pub fn call<F: Fn()>(f: F) { f() }

pub fn main() {
    let a = || { println!("a") };
    let b = || { println!("b") };

    call(a); // a::call::<a::main::{closure#0}> 
    call(b); // a::call::<a::main::{closure#1}>
             //
    call(&a as &dyn Fn()); // a::call::<&dyn core::ops::function::Fn<(), Output = ()>>
    call(&b as &dyn Fn()); // a::call::<&dyn core::ops::function::Fn<(), Output = ()>>
}
```

Reply:
Done.